File: behavioral/typescript/mediator_structured/air_traffic_control.ts
import { Mediator } from './interfaces';

export class Aircraft {
  protected controlTower: Mediator;
  protected flightNumber: string;
  protected altitude: number = 0;
  
  constructor(tower: Mediator, flightNumber: string) {
    this.controlTower = tower;
    this.flightNumber = flightNumber;
  }
  
  requestTakeoff(): void {
    console.log(`${this.flightNumber} requesting takeoff clearance...`);
  }
  
  requestLanding(): void {
    console.log(`${this.flightNumber} requesting landing clearance...`);
  }
  
  setAltitude(alt: number): void {
    this.altitude = alt;
    console.log(`${this.flightNumber} at altitude ${this.altitude} feet`);
  }
  
  getFlightNumber(): string {
    return this.flightNumber;
  }
}

export class ControlTower implements Mediator {
  private aircrafts: Map<string, Aircraft> = new Map();
  
  registerAircraft(aircraft: Aircraft): void {
    this.aircrafts.set(aircraft.getFlightNumber(), aircraft);
  }
  
  sendMessage(message: string, sender: string): void {
    console.log(`Control Tower: ${message} from ${sender}`);
  }
  
  grantTakeoff(flightNumber: string): void {
    console.log(`Control Tower: Takeoff granted to ${flightNumber}`);
    const aircraft = this.aircrafts.get(flightNumber);
    if (aircraft) {
      aircraft.setAltitude(10000);
    }
  }
  
  grantLanding(flightNumber: string): void {
    console.log(`Control Tower: Landing granted to ${flightNumber}`);
    const aircraft = this.aircrafts.get(flightNumber);
    if (aircraft) {
      aircraft.setAltitude(0);
    }
  }
}

File: behavioral/dart/mediator_structured/air_traffic_control.dart
import 'interfaces.dart';

class Aircraft {
  final String _flightNumber;
  int _altitude = 0;
  
  Aircraft(this._flightNumber);
  
  void requestTakeoff() {
    print('$_flightNumber requesting takeoff clearance...');
  }
  
  void requestLanding() {
    print('$_flightNumber requesting landing clearance...');
  }
  
  void setAltitude(int alt) {
    _altitude = alt;
    print('$_flightNumber at altitude $_altitude feet');
  }
  
  String get flightNumber => _flightNumber;
}

class ControlTower implements Mediator {
  final Map<String, Aircraft> _aircrafts = {};
  
  void registerAircraft(Aircraft aircraft) {
    _aircrafts[aircraft.flightNumber] = aircraft;
  }
  
  @override
  void sendMessage(String message, String sender) {
    print('Control Tower: $message from $sender');
  }
  
  void grantTakeoff(String flightNumber) {
    print('Control Tower: Takeoff granted to $flightNumber');
    final aircraft = _aircrafts[flightNumber];
    if (aircraft != null) {
      aircraft.setAltitude(10000);
    }
  }
  
  void grantLanding(String flightNumber) {
    print('Control Tower: Landing granted to $flightNumber');
    final aircraft = _aircrafts[flightNumber];
    if (aircraft != null) {
      aircraft.setAltitude(0);
    }
  }
}
